File: frontend/src/lib/effects/dither.ts
import { LayerConfig } from "../../types";
import { hexToRgb } from "../color";
import { getLuminance } from "../mask";

const DITHER_PATTERNS: Record<string, number[][]> = {
    "Classic 2x2": [
        [0, 128],
        [192, 64]
    ],
    "Fine Grid": [
        [0, 192, 48, 240],
        [128, 64, 176, 112],
        [32, 224, 16, 208],
        [160, 96, 144, 80]
    ]
};

/**
 * Replicates bitboylab_boy.core.effects.dither
 */
export function applyDither(
    sourceImageData: ImageData,
    layer: LayerConfig
): ImageData {
    const size = layer.size !== undefined ? Math.max(1, Number(layer.size)) : 4;
    const colorHex = layer.col as string || "#00ff41";
    const patternType = layer.pattern as string || "Classic 2x2";
    
    const { width, height, data } = sourceImageData;
    const outData = new Uint8ClampedArray(width * height * 4);
    const [rStr, gStr, bStr] = hexToRgb(colorHex);

    // 1. Convert source to grayscale 2D pseudo array representation for error diffusion
    // We scale by block size (s) first if needed, but Python loops in jumps of `s`
    // Wait, Python's dither loops over `y in range(0, h, s)` and `x in range(0, w, s)`.
    // And error jumps by `s`. We need a Float32Array for accumulation.
    
    const gray = new Float32Array(width * height);
    for (let i = 0, j = 0; i < data.length; i += 4, j++) {
        gray[j] = getLuminance(data[i], data[i + 1], data[i + 2]);
    }

    const setBlock = (y: number, x: number, s: number, r: number, g: number, b: number) => {
        for (let dy = 0; dy < s; dy++) {
            const py = y + dy;
            if (py >= height) continue;
            for (let dx = 0; dx < s; dx++) {
                const px = x + dx;
                if (px >= width) continue;
                const idx = (py * width + px) * 4;
                outData[idx] = r;
                outData[idx + 1] = g;
                outData[idx + 2] = b;
                outData[idx + 3] = 255;
            }
        }
    };

    if (patternType === "Floyd-Steinberg") {
        for (let y = 0; y < height; y += size) {
            for (let x = 0; x < width; x += size) {
                const idx = y * width + x;
                const oldP = gray[idx];
                const newP = oldP > 127 ? 255 : 0;
                gray[idx] = newP;
                
                if (newP === 255) {
                    setBlock(y, x, size, rStr, gStr, bStr);
                }
                
                const err = oldP - newP;
                if (x + size < width) gray[y * width + (x + size)] += err * 7 / 16;
                if (y + size < height) {
                    const nextRow = (y + size) * width;
                    if (x - size >= 0) gray[nextRow + (x - size)] += err * 3 / 16;
                    gray[nextRow + x] += err * 5 / 16;
                    if (x + size < width) gray[nextRow + (x + size)] += err * 1 / 16;
                }
            }
        }
    } else if (patternType === "Atkinson") {
        for (let y = 0; y < height; y += size) {
            for (let x = 0; x < width; x += size) {
                const idx = y * width + x;
                const oldP = gray[idx];
                const newP = oldP > 127 ? 255 : 0;
                gray[idx] = newP;

                if (newP === 255) {
                    setBlock(y, x, size, rStr, gStr, bStr);
                }

                const err = (oldP - newP) / 8.0;
                
                const neighbors = [
                    [y, x + size],
                    [y, x + 2 * size],
                    [y + size, x - size],
                    [y + size, x],
                    [y + size, x + size],
                    [y + 2 * size, x]
                ];
                
                for (const [ny, nx] of neighbors) {
                    if (ny >= 0 && ny < height && nx >= 0 && nx < width) {
                        gray[ny * width + nx] += err;
                    }
                }
            }
        }
    } else if (patternType === "Random") {
        // Pseudo noise threshold
        for (let y = 0; y < height; y += size) {
            for (let x = 0; x < width; x += size) {
                const threshold = Math.random() * 256;
                if (gray[y * width + x] > threshold) {
                    setBlock(y, x, size, rStr, gStr, bStr);
                }
            }
        }
    } else if (DITHER_PATTERNS[patternType]) {
        // Ordered Bayer
        const pattern = DITHER_PATTERNS[patternType];
        const ph = pattern.length;
        const pw = pattern[0].length;

        for (let y = 0; y < height; y += size) {
            for (let x = 0; x < width; x += size) {
                const pY = (y / size) % ph;
                const pX = (x / size) % pw;
                const threshold = pattern[pY][pX];
                
                if (gray[y * width + x] > threshold) {
                    setBlock(y, x, size, rStr, gStr, bStr);
                }
            }
        }
    }

    return new ImageData(outData, width, height);
}

File: legacy-python/core/effects/dither.py
"""
bitboylab_boy.core.effects.dither
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Multiple dithering algorithms: ordered patterns, error diffusion, and random noise.

Available patterns
------------------
Floyd-Steinberg : Classic error-diffusion with 4-neighbour spread.
Atkinson        : Mac-era error diffusion with 1/8 divisor and 6 neighbours.
Random          : Noise-threshold dithering (no error diffusion).
Classic 2x2     : Ordered dithering with a 2×2 Bayer matrix.
Fine Grid       : Ordered dithering with a 4×4 Bayer matrix.

Performance notes
-----------------
- Floyd-Steinberg and Atkinson use Python loops over (W/s × H/s) cells.
  For large images with small ``size``, this can be slow.
  Future optimisation: port to Numba @jit or Cython extension.
- Random and pattern-based methods are fully vectorisable and very fast.
"""
from __future__ import annotations

import numpy as np
from PIL import Image

from utils.color import hex_to_rgb

# Pre-defined ordered dither threshold matrices
DITHER_PATTERNS: dict[str, np.ndarray] = {
    "Classic 2x2": np.array([[0, 128], [192, 64]], dtype=np.uint8),
    "Fine Grid": np.array(
        [
            [0, 192, 48, 240],
            [128, 64, 176, 112],
            [32, 224, 16, 208],
            [160, 96, 144, 80],
        ],
        dtype=np.uint8,
    ),
}

ALL_DITHER_PATTERNS: list[str] = (
    list(DITHER_PATTERNS.keys()) + ["Floyd-Steinberg", "Atkinson", "Random"]
)


def apply_dither(
    img: Image.Image,
    mask: np.ndarray,
    size: int,
    color_hex: str,
    pattern_type: str,
) -> Image.Image:
    """Apply the requested dither algorithm to *img* within *mask*.

    Args:
        img:          Source image (any mode).
        mask:         Boolean NumPy array (H, W).
        size:         Cell / block size in pixels (1–200).
        color_hex:    Hex color for the "on" pixels.
        pattern_type: Algorithm name (see module docstring).

    Returns:
        RGB PIL image with dither applied over a black background.
    """
    rgb = hex_to_rgb(color_hex)
    gray = np.array(img.convert("L"))
    h, w = gray.shape
    out = np.zeros((h, w, 3), dtype=np.uint8)
    s = max(1, int(size))

    if pattern_type == "Floyd-Steinberg":
        _floyd_steinberg(gray, mask, out, rgb, s, h, w)

    elif pattern_type == "Atkinson":
        _atkinson(gray, mask, out, rgb, s, h, w)

    elif pattern_type == "Random":
        _random_dither(gray, mask, out, rgb, s, h, w)

    elif pattern_type in DITHER_PATTERNS:
        _ordered(gray, mask, out, rgb, s, h, w, DITHER_PATTERNS[pattern_type])

    return Image.fromarray(out, "RGB")


# ---------------------------------------------------------------------------
# Private algorithm implementations
# ---------------------------------------------------------------------------

def _floyd_steinberg(
    gray: np.ndarray,
    mask: np.ndarray,
    out: np.ndarray,
    rgb: tuple[int, int, int],
    s: int,
    h: int,
    w: int,
) -> None:
    """Floyd-Steinberg error diffusion — modifies *out* in-place."""
    pixels = gray.astype(np.float32)
    for y in range(0, h, s):
        for x in range(0, w, s):
            my, mx = min(y, h - 1), min(x, w - 1)
            if not mask[my, mx]:
                continue
            old_p = pixels[my, mx]
            new_p = 255.0 if old_p > 127 else 0.0
            pixels[my, mx] = new_p
            out[y : y + s, x : x + s] = rgb if new_p == 255 else (0, 0, 0)
            err = old_p - new_p
            # Distribute quantisation error to 4 neighbours (scaled by s)
            if x + s < w:
                pixels[y, x + s] += err * 7 / 16
            if y + s < h:
                if x - s >= 0:
                    pixels[y + s, x - s] += err * 3 / 16
                pixels[y + s, x] += err * 5 / 16
                if x + s < w:
                    pixels[y + s, x + s] += err * 1 / 16


def _atkinson(
    gray: np.ndarray,
    mask: np.ndarray,
    out: np.ndarray,
    rgb: tuple[int, int, int],
    s: int,
    h: int,
    w: int,
) -> None:
    """Atkinson error diffusion — modifies *out* in-place."""
    pixels = gray.astype(np.float32)
    for y in range(0, h, s):
        for x in range(0, w, s):
            my, mx = min(y, h - 1), min(x, w - 1)
            if not mask[my, mx]:
                continue
            old_p = pixels[my, mx]
            new_p = 255.0 if old_p > 127 else 0.0
            pixels[my, mx] = new_p
            out[y : y + s, x : x + s] = rgb if new_p == 255 else (0, 0, 0)
            err = (old_p - new_p) / 8.0
            # 6 neighbours with equal 1/8 weight
            for ny, nx in [
                (y,       x + s),
                (y,       x + 2 * s),
                (y + s,   x - s),
                (y + s,   x),
                (y + s,   x + s),
                (y + 2*s, x),
            ]:
                if 0 <= ny < h and 0 <= nx < w:
                    pixels[ny, nx] += err


def _random_dither(
    gray: np.ndarray,
    mask: np.ndarray,
    out: np.ndarray,
    rgb: tuple[int, int, int],
    s: int,
    h: int,
    w: int,
) -> None:
    """Noise-threshold dithering — fully vectorised, modifies *out* in-place."""
    rng = np.random.default_rng(42)
    noise = rng.integers(0, 256, size=(h, w), dtype=np.uint8)
    for y in range(0, h, s):
        for x in range(0, w, s):
            my, mx = min(y, h - 1), min(x, w - 1)
            if mask[my, mx] and gray[my, mx] > noise[my, mx]:
                out[y : min(y + s, h), x : min(x + s, w)] = rgb


def _ordered(
    gray: np.ndarray,
    mask: np.ndarray,
    out: np.ndarray,
    rgb: tuple[int, int, int],
    s: int,
    h: int,
    w: int,
    thresholds: np.ndarray,
) -> None:
    """Ordered (Bayer matrix) dithering — modifies *out* in-place."""
    ph, pw = thresholds.shape
    for y in range(0, h, s):
        for x in range(0, w, s):
            if mask[min(y, h - 1), min(x, w - 1)]:
                if gray[y, x] > thresholds[(y // s) % ph, (x // s) % pw]:
                    out[y : min(y + s, h), x : min(x + s, w)] = rgb
